### PR TITLE
basic functionality up. Needs to be resized.

### PR DESCRIPTION
**Pull request by kathomas921**
Ok, so as you'll see its very basic at the moment.  I used straight js in the view because I wasn't sure where to put it - I haven't really done the front end of rails before. I imagine this should be changed.  Also, most of the equipment label (heating element, pump, etc) column is cut off.  I have no idea where to search to reset that. I tried to just put it in a div and set the height and width values, but that didn't work. I have also yet to figure out how to appropriately adjust the x-axis so that it shows minutes. Right now the hours are based off of the multiplication in the timeline method in schedule.rb.  When I had it in minutes, the x axis was 0, 30, 0, 30 rather than 0, 30, 60, 90. Minutes seem like a better option, but hours was less confusing for the moment.

Anyway, there are obviously a lot of issues, but since I just started out, I'd appreciate a code review/suggestions/additions from anyone.

@Ohmbrewer/web-team

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -55,6 +55,7 @@ gem 'particlerb'
 gem 'aasm', '4.4.0'
 gem 'closure_tree', '6.0.0'
 
+
```
Comment (kyleoliveira) [inline Gemfile]: If you're not using the chartkick gem, you should probably revert the changes to the Gemfile and the Gemfile.lock file.

File: app/models/schedule.rb [hunk 2]
```diff
@@ -137,4 +137,18 @@ def clean_up_task_jobs
     end
   end
 
+  def timeline
+    timeline = []
+    self.root_task.self_and_descendants.each do |task|
```
Comment (kyleoliveira) [inline app/models/schedule.rb]: You could probably clean this method up a bit if you used #collect instead of #each - then you wouldn't need to initialize timeline or manually return it (just return the result from the #collect block).

File: app/models/task.rb [hunk 3]
```diff
@@ -264,6 +264,41 @@ def do_hold(last_check)
     end
   end
 
+  # Returns if task is the root task or not
+  def root
```
Comment (kyleoliveira) [inline app/models/task.rb]: This is already defined and we do not want to override it. See https://github.com/mceachen/closure_tree#accessing-data

File: app/models/task.rb [hunk 4]
```diff
@@ -294,6 +329,10 @@ def trigger_options
       triggers.map { |t, _| [t.gsub('on_', '').titlecase, t] }
     end
 
+
```
Comment (kyleoliveira) [inline app/models/task.rb]: In general, we try to clean up unnecessary comments and space before accepting a PR onto master. I know this isn't necessarily at that state yet, but just a heads up.

File: db/schema.rb [hunk 5]
```diff
@@ -26,15 +26,15 @@
   end
 
   create_table "delayed_jobs", force: :cascade do |t|
-    t.integer  "priority",               default: 0, null: false
-    t.integer  "attempts",               default: 0, null: false
-    t.text     "handler",                            null: false
+    t.integer  "priority",   default: 0, null: false
```
Comment (kyleoliveira) [inline db/schema.rb]: What's up with all the changes in this file? It looks like all they do is remove field size limits, which doesn't seem necessary given the task at hand...

Comment (kathomas921) [inline db/schema.rb]: This is odd. I didn't change the schema and the current schema on my comp looks like
  create_table "delayed_jobs", force: :cascade do |t|
    t.integer  "priority",   default: 0, null: false
    t.integer  "attempts",   default: 0, null: false
    t.text     "handler",                null: false
    t.text     "last_error"
    t.datetime "run_at"
    t.datetime "locked_at"
    t.datetime "failed_at"
    t.string   "locked_by"
    t.string   "queue"
    t.datetime "created_at"
    t.datetime "updated_at"
  end

File: app/models/task.rb (hunk 3 above)
Comment (kathomas921) [inline app/models/task.rb]: Fair enough, I haven't looked into what all the gems do yet.

File: app/models/schedule.rb [hunk 6]
```diff
@@ -137,4 +137,18 @@ def clean_up_task_jobs
...
+      timeline.push(["#{task.equipment.type.titlecase}", "#{task.type}"[0..-5], task.trigger_start*1000/60, (task.trigger_start + task.duration)*1000/60])
```
Comment (kyleoliveira) [inline app/models/schedule.rb]: You need a check to see if task.equipment.nil? is true. I think in most cases that won't happen due to validation on Schedule, but I get an error on the Recipes Index page because I have two Tasks without no Equipment set. This happens when you've set the Task for a Thermostat or RIMS rather than an individual piece of Equipment.

File: app/models/schedule.rb [hunk 7]
```diff
@@ -137,4 +137,28 @@ def clean_up_task_jobs
     end
   end
 
+  def timeline
+    timeline = []
+    self.root_task.self_and_descendants.each do |task|
+
+      timeline.push([
+        "#{task.id}", 
+        "#{task.type[0..-5]} #{task.equipment.type.titlecase} #{task.equipment.rhizome_eid}",
```
Comment (kyleoliveira) [inline app/models/schedule.rb]: Instead of using #equipment, use #sprout. Otherwise, if the task is assigned to a Thermostat or RIMS #equipment will return nil. As-is, this will break the Schedules index page.

File: app/models/task.rb [hunk 8]
```diff
@@ -264,6 +264,44 @@ def do_hold(last_check)
...
+    return true if self.id == schedule.root_task_id
+    false
+  end
+
+
+  # Returns time (in milliseconds) when task begins
+  def trigger_start
+    return 0 if self.root?
+
+    case self.trigger
```
Comment (kyleoliveira) [inline app/models/task.rb]: Rails enums give you functions that end in ? for confirming whether the enum is in a given state. So, instead of comparing trigger to some strings, it should look like:

``` ruby
case
  when on_started?
    # stuff
  when on_done?
    # other stuff
  else
    # all the other stuff
end
```

File: app/views/schedules/_task_list.html.erb [hunk 9]
```diff
@@ -1,29 +1,54 @@
...
-      </li>
-      <% end %>
-  <% end %>
-</ul>
\ No newline at end of file
+
+
+<html>
+<head>
+  <script type="text/javascript" src="https://www.gstatic.com/charts/loader.js"></script>
```
Comment (kyleoliveira) [inline app/views/schedules/_task_list.html.erb:3]: Before you're done, you should pull the Javascript portions into a file in the app/assets/javascript folder that can get called from here via the assets pipeline.

File: app/views/schedules/_task_list.html.erb [hunk 10]
```diff
@@ -1,29 +1,54 @@
...
+      for (i = 0; i < timeline.length; i++) {
+        var start_date = timeline[i][3];
+        timeline[i][3] = new Date(start_date);
+      }
+      data.addRows(timeline);
+      //need to translate ruby array task.trigger_start to new Date(milliseconds) for this to render correctly
+      var options = {
+        height: 275,
+        gantt: {
+          defaultStartDateMillis: new Date(2015, 3, 28)
```
Comment (kyleoliveira) [inline app/views/schedules/_task_list.html.erb]: Maybe the default Date should be dynamically set to whatever today is?

**Comment by kyleoliveira** [comment]
Since you don't need to make any actual changes to them, could you copy the following files from the master branch (or otherwise take them out of the commit):
- [ ] db/schema.rb
- [ ] app/views/recipes/index.html.erb
- [ ] app/assets/javascripts/application.js
- [ ] Gemfile.lock

You can do that by just copying the files' content from the raw page on GitHub, pasting it into the respective files, and adding another commit with the update.

KT: I made an actual change to application.js to remove turbolinks. I can come back to that eventually if/when I try again at moving js to the pipeline, but for now, turbolinks will load the page before the js is finished so the graph won't appear until the user refreshes the page. So, for now, I took turbolinks out.

File: app/models/schedule.rb [hunk 11]
```diff
@@ -137,4 +137,32 @@ def clean_up_task_jobs
     end
   end
 
+  def timeline
+    self.root_task.self_and_descendants.collect do |task|
+      
+      if task.sprout != nil
```
Comment (kyleoliveira) [inline app/models/schedule.rb]: This is Ruby, so instead of 

``` ruby
if task.sprout != nil
```

do

``` ruby
unless task.sprout.nil?
```

File: app/models/task.rb [hunk 12]
```diff
@@ -264,6 +264,34 @@ def do_hold(last_check)
     end
   end
 
+  def name
```
Comment (kyleoliveira) [inline app/models/task.rb]: I don't think you need this method. We already have a `#type_name_display` method which formats the Task type's name for display that more or less gives the same results.

File: app/models/task.rb [hunk 13]
```diff
@@ -264,6 +264,34 @@ def do_hold(last_check)
...
+    return "Turn On" if self.type == "TurnOnTask"
+    return "Turn Off" if self.type == "TurnOffTask"
+    return "Run" if self.type == "RunThermostatTask"
+  end
+
+  def percent_ramp
```
Comment (kyleoliveira) [inline app/models/task.rb]: Is this (`#percent_ramp`) a debugging function? I don't see where it's actually used...

File: app/models/task.rb (hunk 12 above)
Comment (kathomas921) [inline app/models/task.rb]: I agree they are quite similar. I made my own because the other method returns "TurnOn" without a space (as opposed to "Turn On"), which I thought looked awkward on the chart. Also, if it is applied to a thermostat, the label on the graph would be "RunThermostat Thermostat 4" as opposed to "Run Thermostat 4", so this method got rid of the double Thermostat.

File: app/models/task.rb (hunk 13 above)
Comment (kathomas921) [inline app/models/task.rb]: Oh sorry, I'll get rid of that. I was trying to see if I could incorporate the ramping period into the graph but at the moment, I don't think google graphs has a good feature to show the ramping stage.

File: app/models/task.rb (hunk 12 above)
Comment (kyleoliveira) [inline app/models/task.rb]: `#type_name_display` is only used for the _task_list.html.erb, so it can be modified. However, 
1) It uses #titlecase, so it should already be adding spaces between words and capitalizing appropriately which handles your first problem there. You can test in the Rails Console:

``` bash
2.2.1 :001 > a = "TurnOnTask"
 => "TurnOnTask" 
2.2.1 :002 > a.titlecase
 => "Turn On Task" 

```

2) Each RIMS has a Thermostat component and can be tasked with a RunThermostatTask. As such, while it's a little janky to have "Run Thermostat Thermostat 4", I think it's still appropriate. If we really hate it in the long run it's easier to customize it in the one spot than duplicate the function.

So, I'd say use `#type_name_display` :)

**Comment by kyleoliveira** [comment]
Ok, looks good to go. Squash these all into one appropriately worded commit and we should be ready for merge.

**Comment by kathomas921** [comment]
Continued on PR #87
